src/main/java/frc/trigon/robot/posesources/PoseSource.java

```diff
@@ -0,0 +1,52 @@
+package frc.trigon.robot.posesources;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public interface PoseSource {
+    /**
+     * @return whether there are new results since the last call to this method
+     */
+    default boolean hasNewResult() {
+        if (getLastTimestamp() == getTimestampSeconds()) return false;
+
+        setLastTimestamp(getTimestampSeconds());
+        return hasResults();
+    }
+
+    /**
+     * @return true if the pose source is ready to provide a pose, false otherwise
+     */
+    boolean canUpdate();
+
+    /**
+     * @return true if the pose source has results, false otherwise
+     */
+    boolean hasResults();
+
+    /**
+     * @return the current timestamp in seconds
+     */
+    double getTimestampSeconds();
+
+    /**
+     * @param gyroAngle the robot's current gyro angle, so that it would get the best pose relative to the current angle
+     * @return the robot's best estimated pose
+     */
+    Pose2d getRobotPose(Rotation2d gyroAngle);
```

Comment:
You're assuming that every PoseSource is going to need the gyro's angle in order to give you an estimation, which is a wrong assumption.